### PR TITLE
Do not report exceptions on long running Excel reads

### LINKED ISSUE #11896: `InterruptException`/`ClosedByInterruptException` are still showing up in long running computations
Same steps to reproduce as in https://github.com/enso-org/enso/issues/11882:
Screenshot from 2024-12-17 15-25-25

Messages exchanged with GUI:
Screenshot from 2024-12-17 15-32-44

The messages eventually disappear but they shouldn't show up at all. Instead we should be indicating that computation is ongoing.

### PR DESCRIPTION
**Pull request by hubertp**
### Pull Request Description

This change introduces two modifications:
- `ClosedByInterruptException` is wrapped in `InterruptedException` instead of `RuntimeException`
- when instrumentation encounters `InterruptedException` it bails early

Having `ClosedByInterruptException` wrapped in `RuntimeException` meant that it is being reported as a regular `HostException` in the engine and to the user. Instead it should be treated specially since we know that it is caused by cancelling a long-running job. Since it is a statically checked exception it has to be declared and the information has to be propagated through various lambda constructs (thanks Java!).

The above change alone meant that an error is not reported for `Data.read` nodes but any values dependent on it would still report `No_Such_Method` error when the exception is reported as a value. Hence the early bail out mechanism.
Additional _bonus_: LS will send `PendingInterrupted` to GUI that indicates a slightly longer computation that has been interrupted and will be retried.

Closes #11896 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

### CODE REVIEW

**Comment by hubertp** [comment]
Pending some tests but otherwise good to review.

File: engine/language-server/src/main/scala/org/enso/languageserver/runtime/ContextRegistryProtocol.scala [hunk 1]
```diff
@@ -231,9 +231,15 @@ object ContextRegistryProtocol {
...
       }
 
+      /** Indicates that an expression is pending a computation
+        */
       case class Pending(message: Option[String], progress: Option[Double])
           extends Payload
 
+      /** Indicates that an expression's computation has been interrupted and shall be retried.
+        */
+      case object PendingInterrupted extends Payload
```
Comment (radeusgd) [inline engine/language-server/src/main/scala/org/enso/languageserver/runtime/ContextRegistryProtocol.scala]: Shouldn't this be a `wasInterrupted: Boolean` flag on the `Pending` type instead?

I'm not sure why these need to be separate. In the future, won't we also want to report progress on computation that is retried after interruption?

Comment (hubertp) [reply engine/language-server/src/main/scala/org/enso/languageserver/runtime/ContextRegistryProtocol.scala]: That sounds OK, I have no strong opinion about this.

**Review by radeusgd — APPROVED** [review]
I'm wondering if creating some new unchecked variant of `InterruptedException` wouldn't make things easier?

On the other hand it feels nice that we have a clear indication which methods are more prone to interrupts (although it feels like _any_ method can be interrupted anyway, no?).

Anyway, looks OK.

**Comment by hubertp** [comment]
> I'm wondering if creating some new unchecked variant of `InterruptedException` wouldn't make things easier?
> 
> On the other hand it feels nice that we have a clear indication which methods are more prone to interrupts (although it feels like _any_ method can be interrupted anyway, no?).

It's actually fairly mechanical with a decent IDE. And having checked exceptions is actually pretty beneficial in finding bugs.

File: std-bits/table/src/main/java/org/enso/table/util/FunctionWithException.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * <p>This is a <a href="package-summary.html">functional interface</a> whose functional method is
+ * {@link #apply(Object)}.
+ *
+ * @param <T> the type of the input to the function
+ * @param <R> the type of the result of the function
+ * @param <E> the type of the checked exception
+ */
+@FunctionalInterface
+public interface FunctionWithException<T, R, E extends Throwable> {
```
Comment (4e6) [inline std-bits/table/src/main/java/org/enso/table/util/FunctionWithException.java:18]: We are one step away from inventing monads

**Review by 4e6 — APPROVED** [review]

File: std-bits/table/src/main/java/org/enso/table/util/FunctionWithException.java (hunk 2 above)
Comment (hubertp) [reply std-bits/table/src/main/java/org/enso/table/util/FunctionWithException.java:18]: :laughing:

**Pull request merged**